_posts/2021-04-22-torch-for-optimization/torch-for-optimization.Rmd

```diff
@@ -0,0 +1,640 @@
+    X is:  0.0007130796 -0.0002947929 
+
+    Iteration:  3 
+    ...
+    ...
+
+    Value is:  -0.9997789 
+    Gradient is:  3.565234 5.995832 
+    X is:  0.0002042478 -8.457939e-05 
+
+    Value is:  -0.9998025 
+    Gradient is:  -4.614189 -13.74602 
+    X is:  0.0001822711 -7.553725e-05 
+
+    Iteration:  4 
+    ...
+    ...
+
+    Value is:  -0.9999917 
+    Gradient is:  -382.3041 -921.4625 
+    X is:  -6.320081e-06 2.614706e-06 
+
+    Value is:  -0.9999923 
+    Gradient is:  -134.0946 -321.2681 
+    X is:  -6.921942e-06 2.865841e-06 
+
+    Iteration:  5 
+    ...
+    ...
+
+    Value is:  -0.9999999 
+    Gradient is:  -3446.911 -8320.007 
+    X is:  -7.267168e-08 3.009783e-08 
+
+    Value is:  -0.9999999 
+    Gradient is:  -3419.361 -8253.501 
+    X is:  -7.404627e-08 3.066708e-08 
+
+It's still not perfect, but a lot better.
+
+Finally, let's go one step further. Can we use `torch` for constrained optimization?
+
+### Quadratic penalty for constrained optimization
+
+In constrained optimization, we still search for a minimum, but that minimum can't reside just anywhere: Its location has to fulfill some number of additional conditions. In optimization lingo, it has to be *feasible*.
+
+To illustrate, we stay with the flower function, but add on a constraint: $\mathbf{x}$ has to lie outside a circle of radius $sqrt(2)$, centered at the origin. Formally, this yields the inequality constraint
+
+$$
+2 - {x_1}^2 - {x_2}^2 <= 0
+$$
+
+A way to minimize *flower* and yet, at the same time, honor the constraint is to use a penalty function. With penalty methods, the value to be minimized is a sum of two things: the target function's output and a penalty reflecting potential constraint violation. Use of a *quadratic* *penalty*, for example, results in adding a multiple of the square of the constraint function's output:
+
+```{r}
+# x^2 + y^2 >= 2
+# 2 - x^2 - y^2 <= 0
+constraint <- function(x) 2 - torch_square(torch_norm(x))
+
+# quadratic penalty
+penalty <- function(x) torch_square(torch_max(constraint(x), other = 0))
+
+```
+
+A priori, we can't know how big that multiple has to be to enforce the constraint. Therefore, optimization proceeds iteratively. We start with a small multiplier, $1$, say, and increase it for as long as the constraint is still violated:
+
+```{r}
+penalty_method <- function(f, p, x, k_max, rho = 1, gamma = 2, num_iterations = 1) {
+
+  for (k in 1:k_max) {
+    cat("Starting step: ", k, ", rho = ", rho, "\n")
+
+    minimize(f, p, x, rho, num_iterations)
+
+    cat("Value: ",  as.numeric(f(x)), "\n")
+    cat("X: ",  as.matrix(x), "\n")
+    
+    current_penalty <- as.numeric(p(x))
+    cat("Penalty: ", current_penalty, "\n")
+    if (current_penalty == 0) break
+    
+    rho <- rho * gamma
+  }
+
+}
+```
+
+`Minimize()`, called from `penalty_method()`, follows the usual proceedings, but now it minimizes the sum of the target and up-weighted penalty function outputs:
```

Comment:
```suggestion
`minimize()`, called from `penalty_method()`, follows the usual proceedings, but now it minimizes the sum of the target and up-weighted penalty function outputs:
```

Reply:
When referencing code, always keep the casing! (Since R is case sensitive, we, as writers, have to be so, too!)

Reply:
oh yeah right ... needs some discipline, starting a sentence lower-case, though :-)